### PR TITLE
feat(users): Create terminate 2fa API

### PR DESCRIPTION
**Pull request by Riddhiagrawal001**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Currently completion of 2fa and setting the status of the 2fa status as "SET" is handled by `verify_totp` API. We want to remove this from `verify_totp`, so this PR creates a new API to terminate the 2fa flow.

### Additional Changes

- [X] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
This will close the issue #4730 

## How did you test it?
This can only be tested locally as it requires some changes in the redis.

1. terminate 2fa without skip_two_factor_auth query param
  a. If the keys with `TOTP_{user_id}` or `RC_{user_id}` is not present in redis
  Request
  ```
  curl --location 'http://localhost:8080/user/2fa/terminate' \
  --header 'Authorization: Bearer SPT with purpose as TOTP'
  ```

Response 
```
{
    "error": {
        "type": "invalid_request",
        "message": "Two factor auth required",
        "code": "UR_39"
    }
}
``` 
  b. Add the keys to the redis with prefix as `TOTP_{user_id}` or `RC_{user_id}`

  Request
  ```
  curl --location 'http://localhost:8080/user/2fa/terminate' \
  --header 'Authorization: Bearer SPT with purpose as TOTP'
  ```
  Response 
  This will also set the totp_status for the user as "set"
  ```
  {
      "token": "Bearer token",
      "token_type": "next flow token type"
  }
  ```

2. terminate 2fa with skip_two_factor_auth query param
a. skip_two_factor_auth=true
Irrespective if the keys are present in redis or not if skip_two_factor_auth is sent as true it will not change the status and will give the token and token_type for the next flow
Request
```
curl --location 'http://localhost:8080/user/2fa/terminate?skip_two_factor_auth=true' \
--header 'Authorization: Bearer SPT with purpose as TOTP'
```
Response
```
{
    "token": "Bearer token",
    "token_type": "next flow token type"
}
```

b. skip_two_factor_auth=false
Request
```
curl --location 'http://localhost:8080/user/2fa/terminate?skip_two_factor_auth=false' \
--header 'Authorization: Bearer SPT with purpose as TOTP'
```
Response 
This will also set the totp_status for the user as "set"
  ```
  {
      "token": "Bearer token",
      "token_type": "next flow token type"
  }
  ```

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/user.rs [hunk 1]
```diff
@@ -1766,3 +1768,49 @@ pub async fn generate_recovery_codes(
...
+        .find_user_by_id(&user_token.user_id)
+        .await
+        .change_context(UserErrors::InternalServerError)?
+        .into();
+
+    if skip_2fa.is_none() {
+        if !(check_totp_in_redis(&state, &user_token.user_id).await?
+            || check_access_code_in_redis(&state, &user_token.user_id).await?)
+        {
+            return Err(UserErrors::TotpRequired.into());
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:1788]: Add a new error `2FARequired`.

File: crates/router/src/core/user.rs [hunk 2]
```diff
@@ -1766,3 +1768,49 @@ pub async fn generate_recovery_codes(
...
+    skip_2fa: Option<bool>,
+) -> UserResponse<user_api::TokenResponse> {
+    let user_from_db: domain::UserFromStorage = state
+        .store
+        .find_user_by_id(&user_token.user_id)
+        .await
+        .change_context(UserErrors::InternalServerError)?
+        .into();
+
+    if skip_2fa.is_none() {
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:1784]: Should be `!skip_2fa.unwrap_or(false)`.

File: crates/router/src/core/user.rs [hunk 3]
```diff
@@ -1766,3 +1768,49 @@ pub async fn generate_recovery_codes(
...
+        if !(check_totp_in_redis(&state, &user_token.user_id).await?
+            || check_access_code_in_redis(&state, &user_token.user_id).await?)
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:1786]:
```suggestion
        if !check_totp_in_redis(&state, &user_token.user_id).await?
            && !check_access_code_in_redis(&state, &user_token.user_id).await?
```
This is more readable imo.

File: crates/router/src/core/user.rs (hunk 1 above)
Comment (Riddhiagrawal001) [reply crates/router/src/core/user.rs]: New error added

File: crates/router/src/core/user.rs (hunk 2 above)
Comment (Riddhiagrawal001) [reply crates/router/src/core/user.rs]: Changed

File: crates/router/src/core/user.rs (hunk 3 above)
Comment (Riddhiagrawal001) [reply crates/router/src/core/user.rs]: Changed

**Review by ThisIsMani — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

**Review by apoorvdixit88 — DISMISSED** [review]

File: crates/router/src/core/user.rs [hunk 4]
```diff
@@ -1766,3 +1765,51 @@ pub async fn generate_recovery_codes(
         recovery_codes: recovery_codes.into_inner(),
     }))
 }
+
+pub async fn terminate_two_factor_auth(
+    state: AppState,
+    user_token: auth::UserFromSinglePurposeToken,
+    skip_two_factor_auth: Option<bool>,
```
[the remark this answers is not in the document]
Comment (Riddhiagrawal001) [reply crates/router/src/core/user.rs]: Addressed

**Review by racnan — APPROVED** [review]

**Review by apoorvdixit88 — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

**Pull request merged**